### ISSUE TITLE
Instances‘ meaning

### ISSUE THREAD

**Issue by Eevan-zq** [liangyuRain/ForestColl#1]
Hello Sir, I would like to ask about the   _**instance**_   parameter that appears when defining a custom algorithm in MSCCL-TOOLS. However, I have always been unclear about its meaning.
![image](https://github.com/user-attachments/assets/b9e069eb-3310-4f07-ac11-e2d4b71040bc)

When looking at your code, I noticed the statement   _**for inst in range(ninstance):**_   in the to_xml.py file. I suspect that ninstance represents the number of processes created for each GPU. I am not sure if my understanding is correct.
![image](https://github.com/user-attachments/assets/d5d52bce-dd89-43c2-99a8-ff681ddfb42f)

**Comment by liangyuRain** [comment]
The `ninstance` represents the number of identical communication algorithms you are simultaneously running. In GPU hardware, often you need multiple instances sending/receiving data simultaneously to fully utilize bandwidth.

**Comment by Eevan-zq** [comment]
Thank you for your response. I would like to follow up by asking: what factors determine the number of **_instances_** ? For example, is it proportional to the number of threads per GPU, the number of channels, or something else?

**Comment by liangyuRain** [comment]
Usually, if you aim for throughput (for large data sizes), you want as many instances as possible. If the data size is small, then you dont want a lot of instances. In practice, different data sizes have different optimal ninstance. People usually switch schedules based on data size.

**Comment by Eevan-zq** [comment]
Yes, in my understanding, an _**instance**_ is equivalent to the concept of parallelism, where multiple parallel GPUs execute the collective communication primitive algorithm simultaneously, right?

**Comment by liangyuRain** [comment]
The number of GPUs in collective communication is constant regardless of ninstance. Increasing ninstance is increasing the number of parallel threadblocks or SMs executing the schedule.

**Comment by Eevan-zq** [comment]
Thank you for your response. So, in the XML file called by MSCCL, under each GPU tag, there are multiple thread tags, and the value of `ninstance` is positively correlated with the number of these thread blocks.
